yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,412 @@
+    progression over source string. `enforce_monotonic` enforces monotonic
+    state transition (model progresses over each source character), and
+    `attention_context` allows conditioning of state transition over previous
+    _n_ states. 
+    
+    After:
+        Wu, S. and Cotterell, R. 2019. Exact hard monotonic attention for
+        character-level transduction. In _Proceedings of the 57th Annual
+        Meeting of the Association for Computational Linguistics_, pages
+        1530-1537.
+
+    Original implementation:
+        https://github.com/shijie-wu/neural-transducer
+    """
+
+    enforce_monotonic: bool
+    attention_context: int
+
+    def __init__(
+        self, *args, enforce_monotonic=False, attention_context=0, **kwargs
+    ):
+        """Initializes the encoder-decoder.
+
+        Args:
+            *args: passed to superclass.
+            enforce_monotonic [bool, optional]: Enforce monotonic state
+                transition in decoding.
+            attention_context [int, optional]: Size of context window for
+            conditioning state transition. If 0, state transitions are
+                independent.
+            **kwargs: passed to superclass.
+        """
+        self.enforce_monotonic = enforce_monotonic
+        self.attention_context = attention_context
+        super().__init__(*args, **kwargs)
+        self.classifier = nn.Linear(
+            self.decoder.output_size, self.target_vocab_size
+        )
+
+    def init_decoding(self, encoder_out, encoder_mask):
+        """Initializes hiddens and initial decoding output.
+
+        Simply feeds a BOS string to decoder to provide initial probability.
+        """
+        batch_size, _ = encoder_mask.shape
+        decoder_hiddens = self.init_hiddens(batch_size, self.decoder_layers)
+        bos = (
+            torch.tensor(
+                [self.start_idx], device=self.device, dtype=torch.long
+            )
+            .repeat(batch_size)
+            .unsqueeze(-1)
+        )
+        return self.decode_step(
+            bos, decoder_hiddens, encoder_out, encoder_mask
+        )
+
+    def forward(
+        self,
+        batch: data.PaddedBatch,
+    ) -> torch.Tensor:
+        """Runs the encoder-decoder model.
+
+        The emission tensor returned is of shape 
+        (tgt_len x batch_size x src_len x vocab_size), and the
+        transition tensor returned is of shape
+        (tgt_len x batch_size x src_len x src_len).
+
+        Args:
+            batch (data.PaddedBatch).
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: tensosr of emission
+                 and transition probabilities for each transition
+                 state (i.e., target symbol).
+        """
+        encoder_out = self.source_encoder(batch.source).output
+        if self.has_features_encoder:
+            encoder_features_out = self.features_encoder(batch.features).output
+            # Averages to flatten embedding.
```

Comment:
Ah, good catch. Yeah, they assumed a constant amount of features so I wanted to avoid the projection to avoid some extra code. I've toyed around with this in the past and just adding them seems to do the trick anyhow.